library/src/test/java/com/getstream/sdk/chat/viewmodel/messages/MessageListItemLiveDataTests.kt

```diff
@@ -0,0 +1,122 @@
+        threadMessages = MutableLiveData(listOf())
+        typing = MutableLiveData(listOf())
+        reads = MutableLiveData(listOf())
+
+        sut = MessageListItemLiveData(currentUser, messages, threadMessages, typing, reads)
+    }
+
+    @Test
+    fun `When messages are changed if live data in the thread state should not be observed any values `() {
+        sut.observeForever(observer)
+        threadMessages.value = listOf(createMessage())
+        observer.cleanState()
+
+        messages.value = listOf(createMessage())
+
+        observer.lastObservedValue shouldBeEqualTo null
+    }
+
+    @Test
+    fun `When messages are changed if live data is not in the thread state should observe last assigned value to messages`() {
+        sut.observeForever(observer)
+        val sendingMessage = createMessage()
+
+        messages.value = listOf(sendingMessage)
+
+        observer.lastObservedValue.also { result ->
+            result shouldNotBeEqualTo null
+            result!!.listEntities.any { item ->
+                item is MessageListItem.MessageItem && item.message == sendingMessage
+            } shouldBeEqualTo true
+        }
+    }
+
+    @Test
+    fun `When thread messages are changed should observe last assigned value to thread messages`() {
+        sut.observeForever(observer)
+        val threadMessage = createMessage()
+
+        threadMessages.value = listOf(threadMessage)
+
+        observer.lastObservedValue.also { result ->
+            result shouldNotBeEqualTo null
+            result!!.listEntities.any { item ->
+                item is MessageListItem.MessageItem && item.message == threadMessage
+            } shouldBeEqualTo true
+        }
+    }
+
+    @Test
+    fun `When set thread messages should remove observer from old thread messages and set to new one`() {
+        sut.observeForever(observer)
+        val newThreadMessages = MutableLiveData<List<Message>>(listOf())
+
+        sut.setThreadMessages(newThreadMessages)
+
+        threadMessages.hasObservers() shouldBeEqualTo false
+        newThreadMessages.hasObservers() shouldBeEqualTo true
+    }
+
+    @Test
+    fun `When set thread messages should observe values from new thread messages`() {
+        sut.observeForever(observer)
+        val newThreadMessage = createMessage()
+        val newThreadMessages = MutableLiveData(listOf(newThreadMessage))
+
+        sut.setThreadMessages(newThreadMessages)
+
+        observer.lastObservedValue.let { result ->
+            result shouldNotBeEqualTo null
+            result!!.listEntities.any { item ->
+                item is MessageListItem.MessageItem && item.message == newThreadMessage
+            } shouldBeEqualTo true
+        }
+    }
+}
+
+class TestObserver<T> : Observer<T> {
+    var lastObservedValue: T? = null
+        private set
+
+    override fun onChanged(value: T?) {
+        lastObservedValue = value
+    }
+
+    fun cleanState() {
+        lastObservedValue = null
+    }
+}
```

Comment:
This looks like something we should be reusing in tests, so it could probably live in its own file. My own approach to an Observer like this is [here](https://github.com/rainbowcake/rainbowcake/blob/d1911c0e75d74b2f4f1fc6c9054dca0a37ab77b4/rainbow-cake-test/src/main/java/co/zsmb/rainbowcake/test/MockObserver.kt), and it stores a list of values instead of just the last one, and then things can be asserted on it neatly with extensions.

The `cleanState` method is a good idea though, I don't have that myself. Maybe it could be called `reset`, like a Mockito mock can be reset?

Reply:
Thanks for the feedback! I renamed that method according to Mockito style. Also moved this observer class to a standalone file. In future If we need list of all observed value like in your version we can easily change it